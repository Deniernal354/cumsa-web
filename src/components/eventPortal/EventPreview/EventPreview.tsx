"use client";
import Image from "next/image";
import { useState } from "react";
import {
  IoImage,
  IoCheckmark,
  IoCalendarOutline,
  IoLocationSharp,
  IoCash,
  IoLink,
} from "react-icons/io5";

export type PreviewEvent = {
  title: string;
  date: string;
  location: string;
  fee?: string;
  link?: string;
  image?: string;
  imageStyle?: "cover" | "contain" | "fill" | "scale-down" | "none";
};

export function EventPreview({ event }: { event: PreviewEvent }) {
  const [error, setError] = useState(false);

  // Map CSS object-fit values to Tailwind classes
  const getImageClassName = (style: string | undefined) => {
    switch (style) {
      case "contain":
        return "object-contain";
      case "fill":
        return "object-fill";
      case "scale-down":
        return "object-scale-down";
      case "none":
        return "object-none";
      case "cover":
      default:
        return "object-cover";
    }
  };

  const formatDate = (dateString: string) => {
    if (!dateString || dateString === "YYYY-MM-DD") return "Select Date";
    try {
      const date = new Date(dateString);
      return date.toLocaleDateString("en-US", {
        month: "short",
        day: "numeric",
        year: "numeric",
      });
    } catch {
      return dateString;
    }
  };

  return (
    <div className="group relative bg-white rounded-2xl shadow-lg hover:shadow-xl transition-all duration-300 overflow-hidden transform hover:-translate-y-1 w-60">
      {/* Image Container */}
      <div className="relative aspect-[4/3] overflow-hidden">
        {!error && event.image ? (
          <Image
            src={event.image}
            alt={event.title || "preview"}
            fill
            className={`${getImageClassName(event.imageStyle)} transition-transform duration-300 group-hover:scale-105`}
            onError={() => setError(true)}
          />
        ) : (
          <div className="w-full h-full flex items-center justify-center bg-gradient-to-br from-gray-50 to-gray-100 border-2 border-dashed border-gray-300">
            <div className="text-center p-4">
              <IoImage className="w-8 h-8 text-gray-400 mx-auto mb-2" />
              <p className="text-xs text-gray-500">
                {error ? "Failed to load image" : "No image selected"}
              </p>
            </div>
          </div>
        )}

        {/* Gradient overlay for better text readability */}
        <div className="absolute inset-0 bg-gradient-to-t from-black/70 via-black/20 to-transparent opacity-0 group-hover:opacity-100 transition-opacity duration-300" />

        {/* Preview badge */}
        <div className="absolute top-2 left-2 px-2 py-1 bg-blue-600 text-white text-xs font-medium rounded-full shadow-lg">
          Preview
        </div>

        {/* Registration indicator */}
        {event.link && (
          <div className="absolute top-2 right-2 opacity-0 group-hover:opacity-100 transition-opacity duration-300">
            <div className="w-6 h-6 bg-green-500 rounded-full flex items-center justify-center shadow-lg">
              <IoCheckmark className="w-3 h-3 text-white" />
            </div>
          </div>
        )}
      </div>

      {/* Content */}
      <div className="p-4 space-y-2">
        <h3
          className="font-bold text-gray-900 text-sm leading-tight group-hover:text-blue-600 transition-colors duration-200 overflow-hidden"
          style={{ display: "-webkit-box", WebkitLineClamp: 2, WebkitBoxOrient: "vertical" }}
        >
          {event.title || "Enter Event Title"}
        </h3>

        <div className="space-y-1.5 text-xs text-gray-600">
          <div className="flex items-center gap-2">
            <IoCalendarOutline className="w-3 h-3 text-blue-500 flex-shrink-0" />
            <span className="font-medium">{formatDate(event.date)}</span>
          </div>

          <div className="flex items-center gap-2">
            <IoLocationSharp className="w-3 h-3 text-green-500 flex-shrink-0" />
            <span className="truncate">{event.location || "Enter Location"}</span>
          </div>

          {event.fee && (
            <div className="flex items-center gap-2">
              <IoCash className="w-3 h-3 text-yellow-500 flex-shrink-0" />
              <span className="font-medium text-gray-700">{event.fee}</span>
            </div>
          )}

          {event.link && (
            <div className="flex items-center gap-2">
              <IoLink className="w-3 h-3 text-purple-500 flex-shrink-0" />
              <span className="text-purple-600 font-medium">Registration Link</span>
            </div>
          )}
        </div>
      </div>

      {/* Image style indicator */}
      {event.imageStyle && event.imageStyle !== "cover" && (
        <div className="absolute bottom-2 right-2 px-2 py-1 bg-black/70 text-white text-xs rounded-md opacity-0 group-hover:opacity-100 transition-opacity duration-300">
          {event.imageStyle}
        </div>
      )}
    </div>
  );
}
