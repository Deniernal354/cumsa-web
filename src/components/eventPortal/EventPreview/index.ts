export { EventPreview } from "./EventPreview";
export type { PreviewEvent } from "./EventPreview";
