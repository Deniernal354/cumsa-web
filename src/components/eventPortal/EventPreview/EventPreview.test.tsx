import { render } from "@testing-library/react";
import { EventPreview } from "./EventPreview";

test("EventPreview without image snapshot", () => {
  const event = {
    title: "Test Event",
    date: "2024-01-01",
    location: "Test Location",
  };

  const { asFragment } = render(<EventPreview event={event} />);
  expect(asFragment()).toMatchSnapshot();
});
