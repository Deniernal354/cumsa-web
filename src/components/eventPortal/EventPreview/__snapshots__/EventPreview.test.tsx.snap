// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventPreview without image snapshot 1`] = `
<DocumentFragment>
  <div
    class="group relative bg-white rounded-2xl shadow-lg hover:shadow-xl transition-all duration-300 overflow-hidden transform hover:-translate-y-1 w-60"
  >
    <div
      class="relative aspect-[4/3] overflow-hidden"
    >
      <div
        class="w-full h-full flex items-center justify-center bg-gradient-to-br from-gray-50 to-gray-100 border-2 border-dashed border-gray-300"
      >
        <div
          class="text-center p-4"
        >
          <svg
            class="w-8 h-8 text-gray-400 mx-auto mb-2"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 64H96a64.07 64.07 0 0 0-64 64v256a64.07 64.07 0 0 0 64 64h320a64.07 64.07 0 0 0 64-64V128a64.07 64.07 0 0 0-64-64zm-80 64a48 48 0 1 1-48 48 48.05 48.05 0 0 1 48-48zM96 416a32 32 0 0 1-32-32v-67.63l94.84-84.3a48.06 48.06 0 0 1 65.8 1.9l64.95 64.81L172.37 416zm352-32a32 32 0 0 1-32 32H217.63l121.42-121.42a47.72 47.72 0 0 1 61.64-.16L448 333.84z"
            />
          </svg>
          <p
            class="text-xs text-gray-500"
          >
            No image selected
          </p>
        </div>
      </div>
      <div
        class="absolute inset-0 bg-gradient-to-t from-black/70 via-black/20 to-transparent opacity-0 group-hover:opacity-100 transition-opacity duration-300"
      />
      <div
        class="absolute top-2 left-2 px-2 py-1 bg-blue-600 text-white text-xs font-medium rounded-full shadow-lg"
      >
        Preview
      </div>
    </div>
    <div
      class="p-4 space-y-2"
    >
      <h3
        class="font-bold text-gray-900 text-sm leading-tight group-hover:text-blue-600 transition-colors duration-200 overflow-hidden"
        style="display: -webkit-box; -webkit-line-clamp: 2;"
      >
        Test Event
      </h3>
      <div
        class="space-y-1.5 text-xs text-gray-600"
      >
        <div
          class="flex items-center gap-2"
        >
          <svg
            class="w-3 h-3 text-blue-500 flex-shrink-0"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="none"
              height="384"
              rx="48"
              stroke-linejoin="round"
              stroke-width="32"
              width="416"
              x="48"
              y="80"
            />
            <circle
              cx="296"
              cy="232"
              r="24"
            />
            <circle
              cx="376"
              cy="232"
              r="24"
            />
            <circle
              cx="296"
              cy="312"
              r="24"
            />
            <circle
              cx="376"
              cy="312"
              r="24"
            />
            <circle
              cx="136"
              cy="312"
              r="24"
            />
            <circle
              cx="216"
              cy="312"
              r="24"
            />
            <circle
              cx="136"
              cy="392"
              r="24"
            />
            <circle
              cx="216"
              cy="392"
              r="24"
            />
            <circle
              cx="296"
              cy="392"
              r="24"
            />
            <path
              d="M128 48v32m256-32v32"
              fill="none"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="32"
            />
            <path
              d="M464 160H48"
              fill="none"
              stroke-linejoin="round"
              stroke-width="32"
            />
          </svg>
          <span
            class="font-medium"
          >
            Dec 31, 2023
          </span>
        </div>
        <div
          class="flex items-center gap-2"
        >
          <svg
            class="w-3 h-3 text-green-500 flex-shrink-0"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 32C167.67 32 96 96.51 96 176c0 128 160 304 160 304s160-176 160-304c0-79.49-71.67-144-160-144zm0 224a64 64 0 1 1 64-64 64.07 64.07 0 0 1-64 64z"
            />
          </svg>
          <span
            class="truncate"
          >
            Test Location
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
